teps/0032-tekton-notifications.md

```diff
@@ -0,0 +1,144 @@
+---
+status: proposed
+title: Tekton Notifications
+creation-date: '2020-11-18'
+last-updated: '2020-11-18'
+authors:
+- '@afrittoli'
+---
+
+# TEP-0032: Tekton Notifications
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [User Experience (optional)](#user-experience-optional)
+- [Alternatives](#alternatives)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+## Motivation
+
+Tekton is a tool for building CI/CD systems. It allows automating workflows
+using pipelines which may run autonomously most of the times, until something
+goes wrong or they reach as step that require human involvement.
+
+Notification can help raising the attention of an human being to a specific
+workflow execution that is need of attention.
+
+Several Tekton users ask about notifications in Tekton. We have developed two
+features that may help implementing notifications:
+
+- [Cloud events](https://tekton.dev/docs/pipelines/events/#events-via-cloudevents)
+- [Finally](https://tekton.dev/docs/pipelines/pipelines/#adding-finally-to-the-pipeline)
+
+Cloud events are triggered every time a pipeline or task is started, begins
+executing and is terminated. Cloud events, combined with Tekton Triggers can
+be used to execute tasks and pipelines as a reaction to an event. This is
+used in our dogfooding cluster to generate notifications for failed CD pipelines.
+It works fine, however it requires a good amount of YAML and plumbing resource
+together to achieve the goal of notifications.
+
+The "Finally" section in a pipeline allows execution tasks once all regular tasks
+completed execution. Such "final" tasks can be used to send notifications about
+the outcome of a pipeline, as they will always be invoked, even if a task in the
+main pipeline failed. However final tasks are only executed at the end, so they
+cannot be used to "start" notifications, and they are limited to pipelines.
+Work is in progress to allow "final" tasks access to the execution status of tasks
+in the main pipeline.
+
+Tekton users would highly benefit from native notification support.
+
+### Goals
+
+The main goal of the TEP is to design and implement a notification system that
+is both easy to use and flexible, for delivering notification targeted to humans.
```

Comment:
Also, one more thought here, "easy-to-use" and "flexible" are often at odds with each other when it comes to an implementation.

Could you define how we might strike that balance as part of the goals?  For example, I could imagine something like this:

> ### Goals

> Notifications are easy to use: enabling and configuring default notifications to get something up and running (notifying on every event of every Pipeline/Task) requires only one or two lines of configuration.
>
> Notifications are flexible: users are able to configure the precise Pipelines/Tasks and events that notifications are sent for. Users can also modify the configuration for their chosen notification system.

Reply:
> What differentiates "targeted to humans" and, say, "targeted to machines", from Tekton Pipelines' perspective? Is it about the payload format? In what way could Tekton's native notifications have an effect on how this is used for that?

It may turn-out to be that there is no effect;  I want to keep that in mind as one of the important design principles for this work. If we don't, the risk is that we start fitting in use cases for machine-to-machine triggers, at the cost of usability for users.

Reply:
> Also, one more thought here, "easy-to-use" and "flexible" are often at odds with each other when it comes to an implementation.

Fair point, I will add more details about this.
> 
> Could you define how we might strike that balance as part of the goals? For example, I could imagine something like this:
> 
> > ### Goals
> 
> > Notifications are easy to use: enabling and configuring default notifications to get something up and running (notifying on every event of every Pipeline/Task) requires only one or two lines of configuration.
> > Notifications are flexible: users are able to configure the precise Pipelines/Tasks and events that notifications are sent for. Users can also modify the configuration for their chosen notification system.

Reply:
Again, I like your wording, I will take it as is.
I should add your name to the author list!

Reply:
My 2 cents: I think something important here is leveraging the Tekton catalog - i.e. this is were flexibility comes from. Easy-to-use should come from a sensible (and hopefully standardized) configuration format so deploying new notifiers is easy.

Reply:
Yeah, absolutely. It have "using the catalog" as one of the goals of this TEP in L81-83